### PR TITLE
Improve "toFactory" type safety

### LINKED ISSUE #569: Improve "toFactory" type safety
The following code is incorrect but the compiler is not able to detect the issue due to `any` being used in `toFactory`:

```ts
type WarriorFactory = (name: string) => Warrior;

container.bind<WarriorFactory>("WarriorFactory").toFactory((ctx) => {
    return () => { // Missing name: string
        let ninja = ctx.container.get<Warrior>("Warrior");
        ninja.name = name;
        return ninja;
    }
});
```

### PR DESCRIPTION
**Pull request by leonardssh**
Improvements by @GreenDou in https://github.com/inversify/InversifyJS/issues/569#issuecomment-755968007

Closes https://github.com/inversify/InversifyJS/issues/569

## Description

Improve "toFactory" type safety

## Related Issue

https://github.com/inversify/InversifyJS/issues/569

## Motivation and Context

https://github.com/inversify/InversifyJS/issues/569

## Types of changes

- [ ] Updated docs / Refactor code / Added a tests case (non-breaking change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the changelog.

### CODE REVIEW

File: src/interfaces/interfaces.ts [hunk 1]
```diff
@@ -71,9 +71,9 @@ namespace interfaces {
         cache: null | TActivated | Promise<TActivated>;
     }
 
-    export type Factory<T> = (...args: any[]) => (((...args: any[]) => T) | T);
+    export type Factory<T, U extends any[] = any[]> = (...args: U) => (((...args: U) => T) | T);
```
Comment (notaphplover) [inline src/interfaces/interfaces.ts]:
```suggestion
    export type Factory<T, U extends unknown[] = unknown[]> = (...args: U) => (((...args: U) => T) | T);
```

File: src/interfaces/interfaces.ts [hunk 2]
```diff
@@ -331,7 +331,7 @@ namespace interfaces {
         toConstantValue(value: T): BindingWhenOnSyntax<T>;
         toDynamicValue(func: DynamicValue<T>): BindingInWhenOnSyntax<T>;
         toConstructor<T2>(constructor: Newable<T2>): BindingWhenOnSyntax<T>;
-        toFactory<T2>(factory: FactoryCreator<T2>): BindingWhenOnSyntax<T>;
+        toFactory<T2, T3 extends any[] = any[]>(factory: FactoryCreator<T2, T3>): BindingWhenOnSyntax<T>;
```
Comment (notaphplover) [inline src/interfaces/interfaces.ts]:
```suggestion
        toFactory<T2, T3 extends unknown[] = unknown[]>(factory: FactoryCreator<T2, T3>): BindingWhenOnSyntax<T>;
```

**Review by notaphplover — CHANGES_REQUESTED** [review]
unknown should be better than any imo

**Review by PodaruDragos — APPROVED** [review]

**Review by notaphplover — APPROVED** [review]

**Comment by tonyhallett** [comment]
This pull request has an introduced an error. https://github.com/inversify/InversifyJS/pull/1311#issuecomment-905480952

It has been corrected in https://github.com/inversify/InversifyJS/pull/1311
https://github.com/inversify/InversifyJS/blob/ee0a249c9c33e0dfb050aab41ded25c7473baa84/src/interfaces/interfaces.ts#L78

Also, please check wiki files when changing types.

Note that the correcting pull has not been merged yet.

**Pull request merged**